### PR TITLE
[tt-train] Fix RMSNorm backward computation and improve tests

### PR DESCRIPTION
**Pull request by mdragulaTT**
### Problem description
While running `llama3_7B` training using:

```bash
./build/sources/examples/nano_gpt/nano_gpt \
    -c configs/training_shakespeare_llama3_7B_tensor_parallel.yaml
```

we observed that the **training loss was exploding** after a few steps.  
Investigation showed two main issues in the RMSNorm backward fused kernel:

1. **Incorrect `gamma` broadcasting** – gamma was not being broadcast along rows as required leading to incorrect gradients.  
2. **Missing explicit zeroing** – certain compute ops accumulate into destination registers rather than overwrite them, which left stale values in intermediate buffers.

Both issues resulted in diverging gradients.

### What's changed
- **Fixed gamma broadcasting**: Now correctly broadcast along rows in the backward pass.
- **Added explicit zeroing**: Introduced a dedicated zero-filled circular buffer to reset registers used in intermediate accumulations.
- **Simplified scaling path**: Removed explicit column broadcast and scale CB; replaced with `mul_tiles_bcast_cols`.
- **Improved test coverage**:
  - Tightened error tolerance thresholds.
  - Added a basic correctness test.
  - Added test with NanoGPT-like dimensions to cover real training shapes.
- **Validated fix in real training**:
  - Reran `llama3_7B` training — loss remains stable and decreases over time.
  - See plots in comparison section.

## Comparison: fused vs composite
The main goal was **correctness and training stability**. We compared the *loss curves* of the **composite implementation** vs the **fused implementation** to ensure similar convergence behavior.

Loss comparison plot for **NanoLlama3**:  
image

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/rmsnorm_bw_device_operation.cpp [hunk 1]
```diff
@@ -134,6 +134,8 @@ ttsl::hash::hash_t RMSNormBackwardDeviceOperation::compute_program_hash(
     const auto& input_tensor = tensor_args.input;
     const auto& input_logical_shape = input_tensor.logical_shape();
     auto program_factory = select_program_factory(args, tensor_args);
+    // TODO(maciek): Consider adding gamma shape to the hash as well. It is not strictly necessary, as gamma shape is
+    // derived from input shape, but it would make the hash more robust to changes in the code I guess.
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/rmsnorm_bw_device_operation.cpp]: I disagree. Gamma shape is a function of input tensor shape. 
To make it robust, lets make an explicit check in the appropriate section

File: tt-train/tests/ops/rmsnorm_op_test.cpp [hunk 2]
```diff
@@ -453,12 +457,17 @@ TEST_F(RMSNormOpTest, RMSNorm_Compare_BlockSize2_EvenC) {
...
 }
 
-// Test training-like shapes with realistic model dimensions
+// Test training-like shapes with NanoLlama dimensions
 TEST_F(RMSNormOpTest, RMSNorm_Compare_TrainingShapes_NanoLlama) {
     // NanoLlama training shape: batch=64, seq_len=256, hidden_dim=384
     CompareKernelVsComposite({64U, 1U, 256U, 384U});
 }
 
+// Test training-like shapes with NanoGPT dimensions
```
Comment (rfurko-tt) [inline tt-train/tests/ops/rmsnorm_op_test.cpp]: is it nanogpt shape? seems like LLaMA7B embedding dim :)

File: tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/rmsnorm_bw_program_factory.cpp [hunk 3]
```diff
@@ -147,13 +147,11 @@ bool fits_in_l1_check(
...
     const uint64_t recip_rms_a_bcasted_memory = kNumRecipRmsATiles * bfloat16_single_tile_size_bytes;
     const uint64_t scale_memory = kNumScaleTiles * float32_single_tile_size_bytes;
-    const uint64_t scale_bcasted_memory = kNumScaleBcastedTiles * float32_single_tile_size_bytes;
 
     // Total L1 memory required
     const uint64_t required_L1_in_bytes = input_memory + mask_memory + scaler_memory + gamma_memory + rms_a_memory +
                                           dL_dout_memory + matmul_reduce_memory + dL_da_memory +
-                                          dL_dgamma_components_memory + recip_rms_a_bcasted_memory + scale_memory +
-                                          scale_bcasted_memory;
+                                          dL_dgamma_components_memory + recip_rms_a_bcasted_memory + scale_memory;
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/rmsnorm_bw_program_factory.cpp]: i feel like this calculation missing cb that contains zero? right? Imagine the case when we are exactly at the edge of fitting

Comment (rfurko-tt) [reply tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/rmsnorm_bw_program_factory.cpp]: I feel it's better to be conservative here, otherwise we get OOM

File: tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/kernels/compute/rmsnorm_bw_kernel.cpp [hunk 4]
```diff
@@ -75,10 +82,21 @@ inline void compute_dL_da(
...
+    // NOTE(maciek): We need to manually zero the rhs register, as mul_tiles_bcast_cols accumulates into
+    // the dst register.
+    copy_tile_init(cb_zero_idx);
+    copy_tile(cb_zero_idx, /* tile_idx */ 0, rhs_register);
+
     // Compute scaled_outer = scale * a (broadcasted multiplication)
-    reconfig_data_format(cb_input_idx, cb_scale_bcasted_idx);
-    mul_tiles_init(cb_input_idx, cb_scale_bcasted_idx);
-    mul_tiles(cb_input_idx, cb_scale_bcasted_idx, /* tile_idx */ input_tile_idx, /* tile_idx */ 0, rhs_register);
+    reconfig_data_format(cb_input_idx, cb_scale_idx);
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/kernels/compute/rmsnorm_bw_kernel.cpp:91]: do we need reconfig here? or is it overkill?

**Review by rfurko-tt — APPROVED** [review]

File: tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/kernels/compute/rmsnorm_bw_kernel.cpp (hunk 4 above)
Comment (mdragulaTT) [reply tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/kernels/compute/rmsnorm_bw_kernel.cpp:91]: Yes, we need this because, `cb_scale_idx` is the only CB that has a `precise_data_type`, so we to call reconfig.

File: tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/rmsnorm_bw_device_operation.cpp (hunk 1 above)
Comment (mdragulaTT) [reply tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/rmsnorm_bw_device_operation.cpp]: Done. Added check.

File: tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/rmsnorm_bw_program_factory.cpp (hunk 3 above)
Comment (mdragulaTT) [reply tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/rmsnorm_bw_program_factory.cpp]: Fair enough. Done.

File: tt-train/tests/ops/rmsnorm_op_test.cpp (hunk 2 above)
Comment (mdragulaTT) [reply tt-train/tests/ops/rmsnorm_op_test.cpp]: Yes, ofc! My bad.
